### PR TITLE
[skip ci] Add design doc for the Perf Score feature

### PR DESCRIPTION
**Pull request by briansull**

### CODE REVIEW

**Comment by briansull** [comment]
@dotnet/jit-contrib @mikedn PTAL

Requesting discussion of this new proposed feature to help us measure the performance of the JITted code.

Current implementation is available in PR #26456

File: Documentation/design-docs/Perf-Score.md [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+This current Perf Score feature does **NOT** propose to try to accurately model the actual dynamic execution cost.
+Instead we propose to return an estimate of the dynamic execution cost.  
+We are willing to make several simplfying assumptions that allow us to more quicly and easily return a reasonable estinmate to use as the Perf Score.
+
+### Modern CPU Hardware documentation:
+Hardware vendors, such as Intel and ARM,  provide information for assembly lanaguge programmers and compiler writers.
+They typically provide the following details for instructions:
+   - Latency -- The number of clock cyccles that are required for the execution to complete and produce a result after all inputs are available.
+   - Throughput -- The number of clock cycles required to wait before the issue ports are free to accept another instruction. This value is often less than one cycle.
+   - Pipeline Ports -- Each instruction uses certain pipeline functions, such as LOad Memory, Store Mmeory, Integer ALU, Floating Point Multiply, etc...
```
Comment (sandreenko) [inline Documentation/design-docs/Perf-Score.md]: Nit: LOad -> load, Mmeory -> memory

File: Documentation/design-docs/Perf-Score.md [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+For the Perf Score implementation we will call the method getInsExecutionCharacteristics and it will return the insExecutionCharacteristics for each instruction.
+We will use the information returned along with the BasicBlock weight to calculate the weighted execution cost for each instruction.
+For this implementation we won't try to model the actual def/use latencies and instead will use a simplified model to estimate the instruction latency.
+If in the future we decide to add an instruction scheduler phase to the JIT we can revisit this area.  
+I believe that it will be straightforward to add it at that time.
+
+### Simplfying Assumptions
```
Comment (sandreenko) [inline Documentation/design-docs/Perf-Score.md]: Nit: Simplfying -> Simplifying

File: Documentation/design-docs/Perf-Score.md [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+1. We will assume that we won't have to pay for the latency for memory store operations.  This means that we don't expect a memory read to occur that reads back the value being stored before it has finished it latency cost.
```
Comment (sandreenko) [inline Documentation/design-docs/Perf-Score.md]: Nit: it latency -> its latency

File: Documentation/design-docs/Perf-Score.md [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+2. We will assume that the hardware speculation will be able to hide one cycle of latency for each instruction.
+3. We won't model exact pipeline port usage and instead will model the worst case, such issuing back to back divide intructions.
```
Comment (sandreenko) [inline Documentation/design-docs/Perf-Score.md]: Nit: intructions -> instructions

File: Documentation/design-docs/Perf-Score.md [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+### Additional code size costs
+We also want to take into account the anount of code generated by the JIT for the method.
```
Comment (sandreenko) [inline Documentation/design-docs/Perf-Score.md]: Nit: anount -> amount

File: Documentation/design-docs/Perf-Score.md [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+We do this as follows:
+
+1. We add 0.1 for each byte of hot code generated
```
Comment (sandreenko) [inline Documentation/design-docs/Perf-Score.md]: How do we distinguish hot/cold block when we don't have ngen splitting?

**Comment by CarolEidt** [comment]
> Is it a temporary solution?

I view this as a long-term solution.

> Currently, we don't have many benchmarks in dotnet/performance repo and don't have an easy way to run them (no azp jobs). Are we planning to use the perf score in the long term future?

I don't see the perf score as a replacement for benchmarks, but rather provide a *much* easier way to evaluate changes that may increase perf while also increasing code size. And I expect that it will continue to provide an improved picture of perf impact as compared to jit diff.

File: Documentation/design-docs/Perf-Score.md (hunk 6 above)
Comment (briansull) [reply Documentation/design-docs/Perf-Score.md]: Without hot/cols splitting of code regions, everything will be in the Hot region,
This is always true for crossgen on CoreCLR today.

**Pull request merged**